src/backend/distributed/executor/executor.c

```diff
@@ -1675,6 +1687,37 @@ PlacementAccessListForTask(Task *task, ShardPlacement *taskPlacement)
 		placementAccessList = lappend(placementAccessList, placementAccess);
 	}
 
+	/*
+	 * We've already added anchor shardId's placement access to the list. Now,
+	 * add the other placements in the relationShardList.
+	 */
```

Comment:
I'm a bit confused. Function comments says this function returns placement accesses for the given task and **task placement**, but here we are adding placement accesses for the other shards' placements?

Also aren't we already looping over all tasks and their placements in `AssignTasksToConnections()`, which should cover all shards? Why do we need to do that here again?

Reply:
>  but here we are adding placement accesses for the other shards' placements?

`anchorShardId` is the anchor one 😄  However, a task could access to many placements. For example, `SELECT FROM t1 JOIN t2 USING(id)` accesses two tables, and each task actually accesses two placements, each coming from different tables. So, here `relationShardList ` is used to keep track of these things. And, it makes sense to add all placements to the access list so that the connection/transaction mngmt keeps track of it.

(For example, if you create foreign key, we add the table that is referenced to the `relationShardList` where as the referencing one is on the anchor part.)

> Also aren't we already looping over all tasks and their placements in AssignTasksToConnections(), which should cover all shards? Why do we need to do that here again?

I think we're looping over anchor shards, but we should still loop over `relationShards`

Reply:
I see. It seems that `task->relationShardList` already contains `task->anchorShardId`, so what is that `addAnchorAccess` logic in this function for? I mean the code that you added always seem to add `anchorShardId` access, so we can remove the `if (addAnchorAccess)` part?

Reply:
Ok, it contains it for computeTasks, but not for DDL or DML tasks.

Reply:
There might be some code paths which skips building `relationShardList ` because it is a single table - single shard access. For example, do we set it for single row inserts? I need to do some code reading to make sure that it is OK.

Reply:
> There might be some code paths which skips building relationShardList  because it is a single table - single shard access. For example, do we set it for single row inserts? I need to do some code reading to make sure that it is OK.

It seems that removing `if (addAnchorAccess)` is not Ok, I removed it and tests failed. But this function is overall a bit not easy to understand.

I am still trying to make sure I understand every part of it and then make suggestions, if I have.

Reply:
I think non-foreign-key DDLs do not set `relationShardList `

Reply:
> But this function is overall a bit not easy to understand.

For example, there is no `else if DML ` and we always use `BuildPlacementSelectList ()`. THat's because there cannot be a (distributed) query which modifies multiple tables. Even if a say INSERT .. SELECT, only one table is modified, the others are read. I'll update the comment for this.

Reply:
> I'll update the comment for this.

Ah, sorry. I already have a comment for it:

```
		/*
		 * In case of SELECTs or DML's, we add SELECT placement accesses to the
		 * elements in relationShardList. For SELECT queries, it is trivial, since
		 * the query is literally accesses the relationShardList in the same query.
		 *
		 * For DMLs, create placement accesses for placements that appear in a
		 * subselect.
```